src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using underlying_t = typename std::underlying_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<underlying_t>::max());
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Generates reusable sequential IDs
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class IdRegistry
+{
+    using id_int_t = std::underlying_type_t<ID_T>;
+
+public:
+
+    IdRegistry() = default;
+    IdRegistry(size_t capacity) { reserve(capacity); };
+
+    /**
+     * @brief Create a single ID
+     *
+     * @return A newly created ID
+     */
+    [[nodiscard]] ID_T create()
+    {
+        ID_T output;
+        create(&output, 1);
+        return output;
+    }
+
+    /**
+     * @brief Create multiple IDs
+     *
+     * @param out   [out] Iterator out
+     * @param count [in] Number of IDs to generate
+     */
+    template<typename IT_T>
+    void create(IT_T out, size_t count);
+
+    /**
+     * @return Size required to fit all existing IDs, or allocated size if
+     *         reserved ahead of time
+     */
+    constexpr size_t capacity() const noexcept { return m_deleted.size(); }
+
+    /**
+     * @return Number of registered IDs
+     */
+    constexpr size_t size() const { return capacity() - m_deleted.count(); }
+
+    /**
+     * @brief Allocate space for at least n IDs
+     *
+     * @param n [in] Number of IDs to allocate for
+     */
+    void reserve(size_t n)
+    {
+        m_deleted.resize(n, true);
+    }
+
+    /**
+     * @brief Remove an ID. This will mark it for reuse
+     *
+     * @param id [in] ID to remove
+     */
+    void remove(ID_T id)
+    {
+        if ( ! exists(id))
+        {
+            throw std::runtime_error("ID does not exist");
+        }
+
+        m_deleted.set(id_int_t(id));
+    }
+
+    /**
+     * @brief Check if an ID exists
+     *
+     * @param id [in] ID to check
+     *
+     * @return true if ID exists
+     */
+    bool exists(ID_T id) const noexcept
+    {
+        return ! m_deleted.test(id_int_t(id));
+    };
+
+    template <typename FUNC_T>
+    void for_each(FUNC_T func);
```

Comment:
Wouldn't this be better implemented as a begin() and end() function (returning iterators), and then consumers can use std::for_each if they want?

Reply:
True. At the time I wrote this I just needed a simple for-each.

Reply:
Decided to get rid of it now